### ISSUE TITLE
hledger-ui can crash from keyboard input at startup

### ISSUE THREAD

**Issue by ply** [plaintextaccounting/hledger#743]
Steps to reproduce (I'm using version 1.9):
```
$ hledger ui -f file
<press any key before interface loads (you have to be fast or just very impatient)>
<get error which closes the program, usually segfault>
```

**Comment by simonmichael** [comment]
Wow.. thanks for the report. I can reproduce by `hledger-ui <ENTER><ENTER><ENTER><ENTER>`.

**Comment by Xitian9** [comment]
I'm not able to reproduce this in master. Is this still relevant or can it be closed?

**Comment by simonmichael** [comment]
I can still reproduce this with hledger-1.21, though it's not that easy. I load the 1000 transaction sample journal and hammer on the enter key as fast as I can. I see it die every N attempts, where N is somewhere between 10 and 100. On mac it looks like:
```
...
 123:124:125:126:127:128:129:12a           -149.00 T, -772.31 Z
 123:124:125:126:127:128:129:12a:12b         150.00 T, 650.00 Z
 123:124:125:126:127:128:129:12a:12b:12c   -650.00 A, -150.00 T
────────────────────────────────────────────────────────────────── ?:help t:list/tree -+:depth H:end-bals/changes F:forecast a:add q:quit ──────────────────────────────────────────────────────────────────hledger-ui(40710,0x2016ace00) malloc: *** error for object 0x7fd964e040d0: pointer being freed was not allocated
hledger-ui(40710,0x2016ace00) malloc: *** set a breakpoint in malloc_error_break to debug
Abort trap: 6
$
```
On gnu/linux it looks like:
```
Segmentation fault (core dumped)
$
```

**Comment by simonmichael** [comment]
A much faster way to reproduce it:
```
$ while true; do hledger-ui -f examples/1000x1000x10.journal || break; sleep 0.1; done
```
and hammer on the `q` key until it breaks.

A good next step would be to try this with a minimal brick app.

**Comment by pedro-nonfree** [comment]
Another way to reproduce it: `xdotool type "asdf"; hledger-ui`

**Comment by simonmichael** [comment]
I couldn't reproduce this today on mac. Would appreciate any confirmations that this is still reproducible with hledger-ui 1.43 (all platforms).

**Comment by pedro-nonfree** [comment]
Sorry the delay responding on this, I was using for long  time version `1.31`, unable to find time to upgrade it, I recently upgraded to 1.52.3 and looks like worked fine because **I am no longer able to reproduce it** 🚀 , so it might be safe to close the issue

```
[2026-09-04 Fri 18:35:28] $ hledger --version
hledger 1.52.3, linux-x86_64
[2026-09-04 Fri 18:35:29] $ xdotool type "asdf"; hledger-ui
asdf[2026-09-04 Fri 18:35:31] $ 
```

(I pressed `q` to return to shell)

**Comment by simonmichael** [comment]
Thanks @pedro-nonfree. I suppose you no longer have the old binary, but if you did it would be interesting to know if you can still reproduce with that.

I tried to reproduce just now using [this method](https://github.com/plaintextaccounting/hledger/issues/743#issuecomment-812940791), and could not, with hledger 1.21 or current hledger. (This is a different machine from the one I last tested on.)

**Comment by pedro-nonfree** [comment]
yes, I retain previous versions, because I have an upgrade script (attached [1]) that downloads the latest version and changes the symlinks, so I have the previous hledger versions

```
[2026-09-05 Sat 15:24:19] $ xdotool type "asdf"; hledger-1.23/hledger-ui 
asdfSegmentation fault (core dumped)
[2026-09-05 Sat 15:24:26] $ 
```

```
[2026-09-05 Sat 15:25:07] $ xdotool type "asdf"; hledger-1.31/hledger-ui 
asdfSegmentation fault (core dumped)
[2026-09-05 Sat 15:25:08] $ 
```

I am discovering that it was not 100% reproduceable, looks like there is a "race condition"

``` bash
echo 0 > /tmp/counter
#VERSION=1.23
#VERSION=1.31
VERSION=1.52.3


for i in $(seq 1 10); do
    xdotool type "asdf"; timeout 0.5 hledger-${VERSION}/hledger-ui
    code=$?
    if [ $code -gt 128 ] && [ $code -ne 124 ]; then
        echo $(($(cat /tmp/counter) + 1)) > /tmp/counter
    fi
done

echo "Crashes: $(cat /tmp/counter)"
```

- 1.23
    - first attempt: 7 crashes
    - second attempt: 9 crashes
    - third attempt: 9 crashes
- 1.31
    - first attempt: 10 crashes
    - second attempt: 10 crashes
    - third attempt: 8 crashes
- 1.52.3
    - first attempt: 0 crashes
    - second attempt: 0 crashes


[1] [upgrade.sh](https://github.com/user-attachments/files/31864674/upgrade.sh)

**Comment by simonmichael** [comment]
Interesting! I'm curious, but not enough to dig deeper. That's a clever repro.

**Comment by acinader** [comment]
> Interesting! I'm curious, but not enough to dig deeper. That's a clever repro.

Can we reopen this, please? First read indicates that the  info in this report could be used to create an automated test to reproduce. If so, we can probably figure out what's go on and it could be interesting. Seems worth a (low priority) look. You can assign to me.
